ldelf/ta_elf.c

```diff
@@ -262,6 +265,63 @@ static void check_hashtab(struct ta_elf *elf, void *ptr, size_t num_buckets,
 	check_range(elf, "DT_HASH", ptr, sz);
 }
 
+static void check_gnu_hashtab(struct ta_elf *elf, void *ptr)
+{
+	uint32_t *hashtab = ptr;
+	uint32_t *bucket = NULL;
+	size_t last_sym = 0;
+	/*
+	 * Starting from 4 as the first four words are mandatory and hold
+	 * nbuckets, first symbol index, bloom_size, and bloom shift.
+	 */
+	size_t num_words = 4;
+	size_t sz = 0;
+	size_t i = 0;
+
+	if (!IS_ALIGNED_WITH_TYPE(ptr, uint32_t))
+		err(TEE_ERROR_BAD_FORMAT, "Bad alignment of DT_GNU_HASH %p",
+		    ptr);
+
+	check_range(elf, "DT_GNU_HASH", ptr, num_words);
+
+	/*
+	 * Aside from the four uint32_t's, the structure contains:
+	 * the bloom filter array, which is either an uint32_t or uint64_t
+	 * array of size bloom_size;
+	 * the bucket array, which is an uint32_t array of size nbuckets;
+	 * and the symbols array, which is an uint32_t array.
+	 */
+
+	if (ADD_OVERFLOW(num_words, hashtab[0], &num_words) || /* nbuckets */
+	    ADD_OVERFLOW(num_words, hashtab[2], &num_words) || /* bloom_size */
+	    (!elf->is_32bit &&				/* x2 for uint64_t */
+	     ADD_OVERFLOW(num_words, hashtab[2], &num_words)) ||
+	    MUL_OVERFLOW(num_words, sizeof(uint32_t), &sz))
+		err(TEE_ERROR_BAD_FORMAT, "DT_GNU_HASH overflow");
+
+	/*
+	 * There is no total number of exported symbols in DT_GNU_HASH,
+	 * so we have to find the highest bucket value manually.
+	 * See http://deroko.phearless.org/dt_gnu_hash.txt.
+	 */
+	bucket = hashtab + 4 + hashtab[2] * (elf->is_32bit ? 1 : 2);
```

Comment:
I'd prefer helper variables instead of using `hashtab[0..3]` directly.

Reply:
Even better (I think): I have introduced `struct gnu_hashtab`. Please check again.